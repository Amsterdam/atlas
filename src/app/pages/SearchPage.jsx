import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import SearchContainer from '../containers/SearchContainer';
import MapContainer from '../../map/containers/map/MapContainer';
import QuerySearchContainer from '../containers/QuerySearchContainer';
import { getSearchQuery } from '../../shared/ducks/data-search/selectors';
import { toMapAndPreserveQuery as toMapActionCreator } from '../../store/redux-first-router';
import SplitScreen from '../components/SplitScreen/SplitScreen';

const SearchPage = ({ query, toMap }) => {
  if (query) {
    return <QuerySearchContainer />;
  }
  return (
    <SplitScreen
      leftComponent={(
        <MapContainer isFullscreen={false} toggleFullscreen={toMap} />
      )}
      rightComponent={(
        <SearchContainer />
      )}
    />
  );
};

const mapStateToProps = (state) => ({
  query: getSearchQuery(state)
});

const mapDispatchToProps = (dispatch) => bindActionCreators({
  toMap: toMapActionCreator
}, dispatch);

SearchPage.defaultProps = {
  query: undefined
};

SearchPage.propTypes = {
  query: PropTypes.string,
  toMap: PropTypes.func.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(SearchPage);
