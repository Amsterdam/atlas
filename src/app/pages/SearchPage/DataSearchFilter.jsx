import { Heading, Link, themeColor, themeSpacing } from '@datapunt/asc-ui'
import styled from '@datapunt/asc-core'
import RouterLink from 'redux-first-router-link'
import React from 'react'
import { toDataDetail } from '../../../store/redux-first-router/actions'

const Divider = styled.div`
  width: 200px;
  height: 3px;
  background-color: ${themeColor('secondary')};
  margin: ${themeSpacing(8, 0, 6)};
`

const StyledLink = styled(Link)`
  margin: ${themeSpacing(2, 0)};
`

/**
 * Todo: loading state
 * @param results
 * @returns {*}
 */

export default ({ results }) =>
  results.map(result =>
    result.results && result.results.length ? (
      <div key={result.type}>
        <Divider />
        <Heading $as="h2">{result.label}</Heading>
        <ul>
          {result.results.map(location => (
            <li key={location.id}>
              <StyledLink
                to={toDataDetail([location.id, 'bag', 'nummeraanduiding'])}
                $as={RouterLink}
                variant="with-chevron"
              >
                {location.label}
              </StyledLink>
            </li>
          ))}
        </ul>
      </div>
    ) : null,
  )
