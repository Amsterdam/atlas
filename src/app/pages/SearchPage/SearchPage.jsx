import React, { memo, useEffect, useState } from 'react'
import { useMatomo } from '@datapunt/matomo-tracker-react'
import get from 'lodash.get'
import { clearAllBodyScrollLocks, enableBodyScroll } from 'body-scroll-lock'
import {
  breakpoint,
  Column,
  Container,
  Row,
  Modal,
  themeSpacing,
  TopBar,
  Button,
  Heading,
  Spinner,
  styles,
  constants,
} from '@datapunt/asc-ui'
import styled from '@datapunt/asc-core'
import { Close } from '@datapunt/asc-assets'
import ContentContainer from '../../components/ContentContainer/ContentContainer'
import PageFilterBox from '../../components/PageFilterBox/PageFilterBox'
import SEARCH_PAGE_CONFIG, {
  DEFAULT_LIMIT,
  DATA_FILTERS,
  DATA_SEARCH_PAGES,
  DATASET_SEARCH_PAGES,
  EDITORIAL_SEARCH_PAGES,
} from './config'
import SearchPageResults from './SearchPageResults'
import usePagination from '../../utils/usePagination'
import SearchPageFilters from './SearchPageFilters'
import useCompare from '../../utils/useCompare'
import useSelectors from '../../utils/useSelectors'
import { getActiveFilters, getQuery, getSort } from './SearchPageDucks'
import useDocumentTitle from '../../utils/useDocumentTitle'

const FilterColumn = styled(Column)`
  align-content: flex-start;

  @media screen and ${breakpoint('max-width', 'laptop')} {
    display: none;
  }
`

const StyledModal = styled(Modal)`
  @media screen and ${breakpoint('min-width', 'mobileL')} {
    width: calc(100% - ${themeSpacing(10)});
    max-width: calc(400px + ${themeSpacing(12)});
  }

  ${FilterColumn} {
    top: 0;
    width: 100%;
    position: relative;
    margin-bottom: ${themeSpacing(20)};
    display: block;
    overflow-y: auto;
    left: 0;
    right: ${themeSpacing(5)};

    @media screen and ${breakpoint('max-width', 'mobileL')} {
      ${styles.FilterBoxStyle} {
        border-right: transparent;
        border-left: transparent;
      }
    }
    @media screen and ${breakpoint('min-width', 'mobileL')} {
      padding: 0 ${themeSpacing(5)};
    }
  }

  [role='dialog'] {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    position: relative;
    top: ${constants.HEADER_HEIGHT_SMALL}px;
    height: 100vh;
    max-height: 100%;
    max-width: initial;
    left: 0;
    bottom: 0;
    transform: initial;
  }
`

const ModalCloseButton = styled(Button)`
  margin-left: auto;
`

const ApplyFilters = styled.div`
  position: sticky;
  bottom: ${constants.HEADER_HEIGHT_SMALL}px;
  left: 0;
  z-index: 30; // Todo: implement better z-index strategy
  width: 100%;
  display: flex;
  justify-content: center;
  padding: ${themeSpacing(4)};
  background-color: #fff;
  box-shadow: 0 0 ${themeSpacing(2)} 1px black;
  margin-top: auto;
`

const ApplyFiltersButton = styled(Button)`
  width: 100%;
  display: flex;
  text-align: center;
  justify-content: center;
`

/* TODO: Write tests for the Hooks used in this component */
/* istanbul ignore next */
const SearchPage = ({ isOverviewPage, currentPage }) => {
  const [extraQuery, setExtraQuery] = useState({})
  const [showLoadMore, setShowLoadMore] = useState(false)
  const [showFilter, setShowFilter] = useState(false)
  const [query, sort, activeFilters] = useSelectors([getQuery, getSort, getActiveFilters])
  const from = 0
  const defaultSort = isOverviewPage ? 'date:desc' : ''

  const { documentTitle } = useDocumentTitle()
  const { trackPageView } = useMatomo()

  useEffect(() => {
    if (documentTitle) {
      trackPageView({ documentTitle })
    }
  }, [documentTitle])

  const {
    fetching,
    errors,
    totalCount,
    fetchingMore,
    results,
    filters,
    fetchMore,
    hasMore,
  } = usePagination(
    SEARCH_PAGE_CONFIG[currentPage],
    {
      q: query,
      limit: DEFAULT_LIMIT,
      ...extraQuery,
    },
    sort || defaultSort,
    DEFAULT_LIMIT,
    from,
  )

  const currentPageHasChanged = useCompare(currentPage)

  // Close the filterbox when changing the page
  useEffect(() => {
    if (currentPageHasChanged) {
      setShowFilter(false)
    }
  }, [currentPage])

  // Enable / disable body lock when opening the filter on mobile
  useEffect(() => {
    const action = showFilter || currentPageHasChanged ? clearAllBodyScrollLocks : enableBodyScroll
    action()
  }, [showFilter, currentPage])

  useEffect(() => {
    // Always reset the filterboxes when currentPage or data has changed
    if (DATA_SEARCH_PAGES.includes(currentPage)) {
      const types = get(activeFilters.find(({ type }) => type === DATA_FILTERS), 'values', null)
      setExtraQuery({
        types,
      })

      setShowLoadMore(!!types)
    } else if (
      EDITORIAL_SEARCH_PAGES.includes(currentPage) ||
      DATASET_SEARCH_PAGES.includes(currentPage)
    ) {
      setShowLoadMore(true)
      setExtraQuery({
        filters:
          activeFilters.length > 0
            ? activeFilters.map(({ type, values }) => ({
                type,
                values: Array.isArray(values) ? values : [values],
                multiSelect: Array.isArray(values),
              }))
            : null,
      })
    } else {
      setExtraQuery({})
      setShowLoadMore(false)
    }
  }, [currentPage, activeFilters])

  const Filters = (
    <FilterColumn wrap span={{ small: 0, medium: 0, big: 0, large: 4, xLarge: 3 }}>
      {!isOverviewPage && <PageFilterBox {...{ query, currentPage }} />}
      <SearchPageFilters {...{ currentPage, filters, totalCount, query }} />
    </FilterColumn>
  )

  return (
    <Container>
      <ContentContainer>
        <Row>
          {!showFilter && Filters}
          <StyledModal
            aria-labelledby="filters"
            aria-describedby="filters"
            open={showFilter}
            onClose={() => setShowFilter(false)}
            hideOverFlow={false}
          >
            <TopBar>
              <Heading style={{ flexGrow: 1 }} as="h4">
                <ModalCloseButton
                  variant="blank"
                  type="button"
                  size={30}
                  onClick={() => setShowFilter(false)}
                  icon={<Close />}
                />
              </Heading>
            </TopBar>
            {Filters}
            <ApplyFilters>
              <ApplyFiltersButton
                iconLeft={fetching && <Spinner />}
                onClick={() => setShowFilter(false)}
                variant="primary"
              >
                {!fetching && `${totalCount} resultaten tonen`}
              </ApplyFiltersButton>
            </ApplyFilters>
          </StyledModal>
          <SearchPageResults
            {...{
              query,
              errors,
              fetching,
              totalCount,
              results,
              currentPage,
              isOverviewPage,
              hasMore,
              fetchingMore,
              fetchMore,
              showLoadMore,
              sort,
              setShowFilter,
            }}
          />
        </Row>
      </ContentContainer>
    </Container>
  )
}

export default memo(SearchPage)
