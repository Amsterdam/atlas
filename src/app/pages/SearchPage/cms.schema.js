export default `
cmsSearch(q: $q, input: {limit: $limit, types: $types}) {
  totalCount
  results {
    type
    label
    count
    results {
      id
      label
      type
      date
      slug
      dateLocale
      teaserImage
      coverImage
      link {
        uri
      }
    }
  }
}
`
