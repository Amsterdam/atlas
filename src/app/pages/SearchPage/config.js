import PAGES from '../../pages'
import {
  articleSearchQuery,
  dataSearchQuery,
  datasetSearchQuery,
  publicationSearchQuery,
  specialSearchQuery,
  searchQuery,
} from './graphql.config'

export const MAX_RESULTS = 50

export const DATA_FILTERS = 'dataTypes'

export const QUERY_TYPES = {
  [PAGES.PUBLICATION_SEARCH]: 'publicationSearch',
  [PAGES.ARTICLE_SEARCH]: 'articleSearch',
  [PAGES.SPECIAL_SEARCH]: 'specialSearch',
  [PAGES.DATA_SEARCH]: 'dataSearch',
  [PAGES.DATASET_SEARCH]: 'datasetSearch',
}

export default {
  [PAGES.SEARCH]: {
    resolver: Object.values(QUERY_TYPES),
    query: searchQuery,
    label: 'Alles',
  },
  [PAGES.PUBLICATION_SEARCH]: {
    resolver: QUERY_TYPES[PAGES.PUBLICATION_SEARCH],
    query: publicationSearchQuery,
    label: 'Publicaties',
  },
  [PAGES.ARTICLE_SEARCH]: {
    resolver: QUERY_TYPES[PAGES.ARTICLE_SEARCH],
    query: articleSearchQuery,
    label: 'Artikelen',
  },
  [PAGES.SPECIAL_SEARCH]: {
    resolver: QUERY_TYPES[PAGES.SPECIAL_SEARCH],
    query: specialSearchQuery,
    label: 'In Beeld',
  },
  [PAGES.DATA_SEARCH]: {
    resolver: QUERY_TYPES[PAGES.DATA_SEARCH],
    query: dataSearchQuery,
    label: 'Data',
  },
  [PAGES.DATASET_SEARCH]: {
    resolver: QUERY_TYPES[PAGES.DATASET_SEARCH],
    query: datasetSearchQuery,
    label: 'Datasets',
  },
}
