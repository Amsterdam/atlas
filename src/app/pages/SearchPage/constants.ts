export enum SearchType {
  Search = 'search',
  Data = 'data',
  Dataset = 'dataset',
  Map = 'map',
}

export default SearchType
