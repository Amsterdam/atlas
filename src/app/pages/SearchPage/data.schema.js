const GET_ALL_DATA = `query($q: String!, $limit: Int, $types: [String!]) {
            dataSearch(q: $q, limit: $limit, types: $types) {
              totalCount
              results {
                type
                label
                count
                results {
                  id
                  label
                  type
                }
              }
            }
          }`

const GET_INITIAL_DATA = `query($q: String!, $limit: Int, $types: [String!]) {
            dataSearch(q: $q, limit: $limit, types: $types) {
              totalCount
              results {
                type
                label
                count
                results {
                  id
                  label
                  type
                }
              }
            }
          }`

export default {
  GET_ALL_DATA,
  GET_INITIAL_DATA,
}
