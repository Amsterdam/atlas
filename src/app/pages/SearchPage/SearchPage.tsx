import { Container, Row } from '@amsterdam/asc-ui'
import { useMatomo } from '@datapunt/matomo-tracker-react'
import { clearAllBodyScrollLocks, enableBodyScroll } from 'body-scroll-lock'
import { useEffect, useState } from 'react'
import ContentContainer from '../../components/ContentContainer/ContentContainer'
import { isAllResultsPage, isDataSearchPage, isMapSearchPage } from '../../pages'
import useCompare from '../../utils/useCompare'
import useCurrentPage from '../../utils/useCurrentPage'
import useDocumentTitle from '../../utils/useDocumentTitle'
import useParam from '../../utils/useParam'
import SEARCH_PAGE_CONFIG, { DEFAULT_LIMIT } from './config'
import {
  ActiveFilter,
  activeFiltersParam,
  pageParam,
  queryParam,
  Sort,
  SortOrder,
  sortParam,
} from './query-params'
import SearchPageFilters from './SearchPageFilters'
import SearchPageResults from './SearchPageResults'
import usePagination from './usePagination'

const SearchPage = () => {
  const [activeFilters] = useParam(activeFiltersParam)
  const [page] = useParam(pageParam)
  const [query] = useParam(queryParam)
  const [sort] = useParam(sortParam)
  const [initialLoading, setInitialLoading] = useState(true)
  const [showFilter, setShowFilter] = useState(false)
  const currentPage = useCurrentPage()

  const hasQuery = query.trim().length > 0
  const defaultSort: Sort | null = !hasQuery ? { field: 'date', order: SortOrder.Descending } : null
  const withPagination = isPaginatable(currentPage, activeFilters)

  const { documentTitle } = useDocumentTitle()
  const { trackPageView } = useMatomo()

  useEffect(() => {
    if (documentTitle) {
      trackPageView({ documentTitle })
    }
  }, [documentTitle])

  const { fetching, errors, totalCount, filters, results, pageInfo } = usePagination(
    SEARCH_PAGE_CONFIG[currentPage]?.query,
    {
      q: query,
      page: withPagination ? page : null, // In case the pagination doesn't gets deleted when changing routes
      sort: sort ?? defaultSort,
      limit: !withPagination ? DEFAULT_LIMIT : null,
      withPagination, // Without this no PageInfo will be returned, so the CompactPager won't be rendered
      filters: activeFilters,
    },
    SEARCH_PAGE_CONFIG[currentPage]?.resolver,
  )

  const currentPageHasChanged = useCompare(currentPage)

  // Enable / disable body lock when opening the filter on mobile
  useEffect(() => {
    const action = showFilter || currentPageHasChanged ? clearAllBodyScrollLocks : enableBodyScroll
    action(document.body)
  }, [showFilter, currentPage])

  // Only the initial loading state should render the skeleton components, this prevents unwanted flickering when changing query variables
  useEffect(() => {
    if (currentPageHasChanged) {
      // If the page changes, the skeleton components must be rendered, unless we already have results
      setInitialLoading(!results.length)
    } else if (!!results && !fetching) {
      setInitialLoading(false)
    }
  }, [currentPage, results, fetching])

  return (
    <Container>
      <ContentContainer>
        <Row>
          <SearchPageFilters
            filters={filters}
            totalCount={totalCount}
            hideCount={!isDataSearchPage(currentPage)}
            setShowFilter={setShowFilter}
            query={query}
            currentPage={currentPage}
            fetching={fetching}
            showFilter={showFilter}
          />

          <SearchPageResults
            hasQuery={hasQuery}
            query={query}
            errors={errors}
            loading={initialLoading}
            totalCount={totalCount}
            results={results}
            currentPage={currentPage}
            isOverviewPage={isAllResultsPage(currentPage)}
            page={page}
            setShowFilter={setShowFilter}
            pageInfo={pageInfo}
          />
        </Row>
      </ContentContainer>
    </Container>
  )
}

/**
 * Determines if a search page can be paginated or not.
 *
 * @param currentPage The currently active page.
 * @param activeFilters The currently active filter.
 */
function isPaginatable(currentPage: string, activeFilters: ActiveFilter[]) {
  // The data and map search pages can only be paginated if a subset of data is selected.
  if (isDataSearchPage(currentPage) || isMapSearchPage(currentPage)) {
    return activeFilters.length > 0
  }

  // Every other page can be paginated, besides the 'all results' page, since it has mixed content.
  return !isAllResultsPage(currentPage)
}

export default SearchPage
