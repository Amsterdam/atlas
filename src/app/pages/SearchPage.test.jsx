import React from 'react';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';

import { getDataSearchLocation } from '../../shared/ducks/data-search/selectors';
import SearchPage from './SearchPage';

jest.mock('../../shared/ducks/data-search/selectors');

const initialState = {
};

describe('SearchPage', () => {
  it('shows map search results if no query is provided', () => {
    getDataSearchLocation.mockReturnValue([123, 321]);
    const store = configureMockStore()({ ...initialState });
    const component = shallow(<SearchPage />, { context: { store } }).dive();
    expect(component).toMatchSnapshot();
  });

  it('shows text search result if query is provided', () => {
    getDataSearchLocation.mockReturnValue(undefined);
    const store = configureMockStore()({ ...initialState });
    const component = shallow(<SearchPage />, { context: { store } }).dive();
    expect(component).toMatchSnapshot();
  });
});
