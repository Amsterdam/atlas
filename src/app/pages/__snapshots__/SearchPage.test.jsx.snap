// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage shows map search results if no query is provided 1`] = `
<SplitScreen
  leftComponent={
    <Connect(MapContainer)
      isFullscreen={false}
      toggleFullscreen={[Function]}
    />
  }
  rightComponent={<Connect(LocationSearch) />}
/>
`;

exports[`SearchPage shows text search result if query is provided 1`] = `<Connect(QuerySearchContainer) />`;
